homeassistant/components/azure_event_hub/__init__.py

```diff
@@ -1,88 +1,260 @@
+        else:
+            entities_filter = generate_filter(**entities_filter)
+    if config.get(CONF_EVENT_HUB_CON_STRING, None):
+        client_args = {"conn_str": config[CONF_EVENT_HUB_CON_STRING]}
+        conn_str_client = True
+    else:
+        client_args = {
+            "host": f"{config[CONF_EVENT_HUB_NAMESPACE]}.servicebus.windows.net",
+            "credential": EventHubSharedKeyCredential(
+                policy=config[CONF_EVENT_HUB_SAS_POLICY],
+                key=config[CONF_EVENT_HUB_SAS_KEY],
+            ),
+            "event_hub_path": config[CONF_EVENT_HUB_INSTANCE_NAME],
+        }
+        conn_str_client = False
+
+    instance = hass.data[DOMAIN] = AEHThread(
+        hass, client_args, conn_str_client, entities_filter
     )
-    async_sender = client.add_async_sender()
-    await client.run_async()
+    instance.async_initialize()
+    instance.start()
+
+    return await instance.async_ready
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    _LOGGER.info("    Shutting down.")
+    instance = hass.data[DOMAIN]
+    instance.queue.put(None)
+    instance.join()
+    return True
 
-    encoder = JSONEncoder()
 
-    async def async_send_to_event_hub(event: Event):
+class AEHThread(threading.Thread):
+    """A threaded event handler class."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client_args: Dict[str, Any],
+        conn_str_client: bool,
+        entities_filter: vol.Schema,
+    ):
+        """Initialize the listener."""
+        threading.Thread.__init__(self, name="AzureEventHub")
+        self.hass = hass
+        self.queue = queue.Queue()
+        self._client_args = client_args
+        self._conn_str_client = conn_str_client
+        self._entities_filter = entities_filter
+        self.async_ready = asyncio.Future()
+        self.write_errors = 0
+        self.shutdown = False
+        # surpress the INFO and below logging on the underlying packages, they are very verbose, even at INFO
+        logging.getLogger("uamqp").setLevel(logging.WARNING)
+        logging.getLogger("azure.eventhub.client_abstract").setLevel(logging.WARNING)
+
+        _LOGGER.info("    Initialized AEH Thread.")
+
+    def event_to_filtered_event_data(self, event: Event):
         """Send states to Event Hub."""
         state = event.data.get("new_state")
         if (
             state is None
             or state.state in (STATE_UNKNOWN, "", STATE_UNAVAILABLE)
-            or not entities_filter(state.entity_id)
+            or not self._entities_filter(state.entity_id)
         ):
             return
+        event_data = EventData(json.dumps(obj=state, cls=JSONEncoder).encode("utf-8"))
+        return event_data
+
+    @callback
+    def _event_listener(self, event: Event):
+        """Listen for new messages on the bus and queue them for AEH."""
+        item = (time.monotonic(), event)
+        self.queue.put(item)
```

Comment:
You can't write from a asyncio loop into a normal queue. I mixup the asyncio / thread a bit too much :) Decide you for one direction to work with threads or coroutine/jobs and use only async commands or sync commands. If you API only support sync calls, use threads otherwise asyncio jobs.

Reply:
thanks for having a look @pvizeli, I took most of the code around the setup for a thread from the Recorder component. I know that doesn't mean the principles used then still apply but I think now in this component, the functions surrounding setup and shutdown are async, while the rest of the running thread is sync, which kind of makes sense to me, especially in combination with a config flow (not blocking everything while setting up the Thread).

Reply:
If you don't mark this function as async with callback, it works to push the event to queue